src/formula_engine.rs

```diff
@@ -0,0 +1,53 @@
+// License: MIT
+// Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+use std::fmt::Debug;
+use std::{
+    ops::{Add, Div, Mul, Neg, Sub},
+    str::FromStr,
+};
+
+use pest::{iterators::Pairs, Parser};
+
+use crate::{
+    error::FormulaError,
+    expression::Expr,
+    parser::{FormulaParser, Rule},
+};
+
+/// FormulaEngine holds the parsed expression and can calculate the result
+/// based on the provided component values.
+#[derive(Debug)]
+pub struct FormulaEngine<T> {
+    expr: Expr<T>,
+}
+
+impl<
+        'a,
+        T: FromStr
+            + Copy
+            + Neg<Output = T>
+            + Add<Output = T>
+            + Sub<Output = T>
+            + Mul<Output = T>
+            + Div<Output = T>
+            + PartialOrd,
+    > FormulaEngine<T>
+where
+    Expr<T>: TryFrom<Pairs<'a, Rule>>,
+    <T as FromStr>::Err: Debug,
+    FormulaError: From<<Expr<T> as TryFrom<Pairs<'a, Rule>>>::Error>,
+{
+    /// Create a new FormulaEngine from a formula string.
+    pub fn try_new(s: &'a str) -> Result<Self, FormulaError> {
```

Comment:
I think the return type should be `Result<(Self, HashSet<usize>), ...>`, where it returns a formula engine and a set of component IDs for which it needs data.

Reply:
@shsms Is there a use-case in which the component set is already known? If so, I would be a bit hesitant to return both `Self` and `HashSet` from the `try_new` method. How about a special getter for the `HashSet` that can be called after you created the `FormulaEngine`.

Reply:
It is tricky.  I also considered a separate method, but that's not as obvious as returning it from the constructor.  When returning from the constructor, users are forced to deal with it and actively ignore it if they don't want it.  But that should never be the case because the formulas come from somewhere else, and without the component IDs, we won't know which resamplers to start.

You can have example lines that users can copy-paste, saying:

```rust
let (engine, component_ids) = FormulaEngine::try_new(...)?;
```

Anyway, later we can focus more on improving the interface.  I think it should be possible to put more orchestration into the FormulaEngine itself and eliminate the need for a HashMap for passing data, for example.  Maybe after we get the evaluation stuff done.

So I'm ok with either approach for now.